File: static/code/tools/c/enumerations.c
#include <stdio.h>
#include <stdlib.h>

// an enumerated type is a type whose value is a value of its underlying type which includes the values of explicitly named enumeration constants
enum color {
  RED,
  BLUE,
  GREEN
};

enum letter { A, B, C = 10, D, E = 1, F, G = F + C };

enum day {
  MONDAY,
  TUESDAY,
  WEDNESDAY,
  THURSDAY = 5,
  FRIDAY = 7,
  SATURDAY,
  SUNDAY
};

// notice that a struct or union does not introduce scope
struct element {
  int n;
  enum state { SOLID, LIQUID, GAS, PLASMA } s;
};

int main(void) {
  // each enumeration constant becomes an integer constant with type int in the enclosing scope and can be used whenever integer constants are required
  enum color r = RED;

  // the value of the first enumeration constant is zero and thereafter the value of each enumeration constant is one more than the previous enumeration constant unless the value is explicitly set
  printf("%d\n", A);  // 0
  printf("%d\n", B);  // 1
  printf("%d\n", C);  // 10
  printf("%d\n", D);  // 11
  printf("%d\n", E);  // 1
  printf("%d\n", F);  // 2
  printf("%d\n", G);  // 12

  // enumerations do not introduce new scope
  enum day d = MONDAY;

  switch (d) {
    case SATURDAY:
    case SUNDAY:
      printf("weekend!");
      break;
    default:
      printf("weekday");
  }

  struct element oxygen;
  oxygen.n = 8;
  oxygen.s = GAS;

  return 0;
}

File: static/code/tools/java/Enumerations.java

// an enumeration is a type that has a fixed set of constant values
enum AlertState {

  LOW("green"),
  MEDIUM("orange"),
  HIGH("red"),
  UNKNOWN;

  // enumerations can have fields, constructors and methods
  private final String color;

  Level(String color) {
    this.color = color;
  }

  public String getColor() {
    return this.color;
  }

  public static Level findByColor(String color) {
    for (Level value: values()) {
      if (color.equals(value.color)) {
        return value;
      }
    }
    return null;
  }
}


Level[] levels = Level.values();
System.out.println(Arrays.toString(levels));  // 

Level low = Level.LOW;
System.out.println(low.name());     // 
System.out.println(low.ordinal());  //

Level medium = Level.valueOf("MEDIUM");
System.out.println(medium.equals(Level.MEDIUM);  // true
System.out.println(medium == Level.MEDIUM);      // true

Level high = Level.HIGH;
System.out.println(high.getColor());           // 
System.out.println(Level.findByColor("red"));  // LOW


// enumerations can be used with switch statements
Level level = Level.HIGH;

switch (level) {
  case LOW:
    System.out.println(LOW.getColor());)]
    break;
  case MEDIUM:
    System.out.println(LOW.getColor());)]
    break;
  case HIGH:
    System.out.println(LOW.getColor());)]
    break;
}




// enumerations may not be generic and cannot be subclassed


// 
// enum Direction {
//   MONDAY,
//   TUESDAY,
//   WEDNESDAY,
//   THURSDAY,
//   FRIDAY,
//   SATURDAY,
//   SUNDAY;
// }

// Day day = Day.WEDNESDAY;

// switch (day) {
//   case MONDAY: case FRIDAY: case SUNDAY:
//     System.out.println("6");
//     break;
//   case TUESDAY:
//     System.out.println("7");
//     break;
//   case THURSDAY: case SATURDAY:
//     System.out.println("8");
//     break;
//   case WEDNESDAY:
//     System.out.println("9");
//     break;
//   default:
//     throw new UnexpectedDayException(day);
// }